File: src/main/java/net/obesephoenix/koalakraftcrystals/crystals/Crystal.java
package net.obesephoenix.koalakraftcrystals.crystals;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import com.mojang.authlib.GameProfile;
import com.mojang.authlib.properties.Property;
import net.minecraft.util.Tuple;
import net.obesephoenix.koalakraftcrystals.KoalaKraftCrystals;
import org.apache.commons.lang.WordUtils;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.block.Biome;
import org.bukkit.craftbukkit.libs.org.codehaus.plexus.util.Base64;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.SkullMeta;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.potion.PotionEffectType;

import javax.naming.Name;

public abstract class Crystal {

    private final String name;
    private final String id;
    private final String textureURL;
    private final Biome biome;
    private final Map<PotionEffectType, Integer> effects;
    private final ChatColor primaryColor;
    private final ChatColor secondaryColor;

    public Crystal(String name, String id, String textureURL, Biome biome, Map<PotionEffectType, Integer> effects, ChatColor primaryColor, ChatColor secondaryColor) {
        this.name = name;
        this.id = id;
        this.textureURL = textureURL;
        this.effects = effects;
        this.biome = biome;
        this.primaryColor = primaryColor;
        this.secondaryColor = secondaryColor;
    }

    protected List<String> addLorePrefix(List<String> lore) { return lore; }
    protected List<String> addLoreSuffix(List<String> lore) {
        return lore;
    }

    protected List<String> generateDefaultLore() {
        List<String> lore = new ArrayList<>();
        lore = this.addLorePrefix(lore);

        lore.add(" ");
        lore.add(ChatColor.GRAY + "Holding this gem in the " + this.getSecondaryColor() + this.getBiome().toString().toLowerCase());
        lore.add(ChatColor.GRAY + "grants the bearer the power of - ");
        lore.add(" ");

        List<String> finalLore = new ArrayList<>();
        effects.forEach((effect, level) -> finalLore.add(this.getPrimaryColor() + "• " +
                WordUtils.capitalize(effect.getName().toLowerCase().replace('_', ' ')) + " " +
                level.toString().replace('1', ' ')));
        lore.addAll(finalLore);

        lore.add(" ");

        lore = this.addLoreSuffix(lore);
        return lore;
    }

    public ItemStack createItemStack() {
        ItemStack item = new ItemStack(Material.PLAYER_HEAD);
        SkullMeta meta = (SkullMeta) item.getItemMeta();
        GameProfile profile = new GameProfile(UUID.randomUUID(),  null);

        byte[] encodedData = Base64.encodeBase64(String.format("{textures:{SKIN:{url:\"%s\"}}}", this.textureURL).getBytes());
        profile.getProperties().put("textures", new Property("textures", new String(encodedData)));

        Field profileField = null;
        try {
            assert meta != null;
            profileField = meta.getClass().getDeclaredField("profile");
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        }

        if (profileField != null) {
            profileField.setAccessible(true);
            try {
                profileField.set(meta, profile);
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        }

        meta.setLore(this.generateDefaultLore());
        meta.getPersistentDataContainer().set(new NamespacedKey(KoalaKraftCrystals.instance, "uuid")
                , PersistentDataType.STRING, UUID.randomUUID().toString());
        meta.setDisplayName(this.getPrimaryColor().toString() + ChatColor.BOLD + WordUtils.capitalize(this.getName()));

        item.setItemMeta(meta);
        return item;
    }

    public boolean inCorrectBiome(Player player) {
        return player.getLocation().getBlock().getBiome() == biome;
    }

    public String getName() {return name;}
    public String getID() {return id;}
    public String getTextureURL() {return textureURL;}
    public Biome getBiome() {return biome;}
    public Map<PotionEffectType, Integer> getEffects() {return effects;}
    public ChatColor getPrimaryColor() {return primaryColor;}
    public ChatColor getSecondaryColor() {return secondaryColor;}

}

File: src/main/java/net/obesephoenix/koalakraftcrystals/crystals/Crystal.kt
package net.obesephoenix.koalakraftcrystals.crystals

import org.bukkit.entity.Player
import org.bukkit.inventory.meta.SkullMeta
import com.mojang.authlib.GameProfile
import com.mojang.authlib.properties.Property
import java.lang.NoSuchFieldException
import java.lang.IllegalAccessException
import org.bukkit.NamespacedKey
import net.obesephoenix.koalakraftcrystals.KoalaKraftCrystals
import org.apache.commons.lang.WordUtils
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.craftbukkit.libs.org.codehaus.plexus.util.Base64
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType
import java.lang.reflect.Field
import java.util.*
import java.util.function.Consumer

abstract class Crystal(
    val name: String,
    val id: String,
    val textureURL: String,
    biome: String,
    effects: List<String>?,
    primaryColor: ChatColor,
    secondaryColor: ChatColor
) {
    val biome: String
    private val effects: MutableList<String>
    val primaryColor: ChatColor
    val secondaryColor: ChatColor

    abstract fun grantEffects(player: Player?): Boolean

    protected open fun addLorePrefix(lore: MutableList<String?>): MutableList<String?> {
        return lore
    }

    protected fun addLoreSuffix(lore: MutableList<String?>): MutableList<String?> {
        return lore
    }

    private fun generateDefaultLore(): List<String?> {
        var lore: MutableList<String?> = ArrayList()
        lore = addLorePrefix(lore)
        lore.add(" ")
        lore.add(ChatColor.GRAY.toString() + "Holding this gem in the " + secondaryColor + biome.lowercase(Locale.getDefault()))
        lore.add(ChatColor.GRAY.toString() + "grants the bearer the power of - ")
        lore.add(" ")
        val finalLore: MutableList<String?> = ArrayList()
        effects.forEach(Consumer { effect: String ->
            finalLore.add(
                primaryColor.toString() + "• " +
                        WordUtils.capitalize(effect.lowercase(Locale.getDefault()))
            )
        })
        lore.addAll(finalLore)
        lore.add(" ")
        lore = addLoreSuffix(lore)
        return lore
    }

    fun createItemStack(): ItemStack {
        val item = ItemStack(Material.PLAYER_HEAD)
        val meta = item.itemMeta as SkullMeta?
        val profile = GameProfile(UUID.randomUUID(), null)
        val encodedData = Base64.encodeBase64(String.format("{textures:{SKIN:{url:\"%s\"}}}", textureURL).toByteArray())
        profile.properties.put("textures", Property("textures", String(encodedData)))
        var profileField: Field? = null
        try {
            assert(meta != null)
            profileField = meta!!.javaClass.getDeclaredField("profile")
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        }
        if (profileField != null) {
            profileField.isAccessible = true
            try {
                profileField[meta] = profile
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }
        }
        meta!!.lore = generateDefaultLore()
        meta.persistentDataContainer.set(
            NamespacedKey(KoalaKraftCrystals.instance, "uuid"), PersistentDataType.STRING, UUID.randomUUID().toString()
        )
        meta.persistentDataContainer.set(
            NamespacedKey(KoalaKraftCrystals.instance, "crystal_id"), PersistentDataType.STRING, id
        )
        meta.setDisplayName(
            primaryColor.toString() + ChatColor.BOLD + WordUtils.capitalize(
                name
            )
        )
        meta.isUnbreakable = true
        item.itemMeta = meta
        return item
    }

    fun getEffects(): List<String> {
        return effects
    }

    init {
        this.effects = ArrayList()
        this.biome = biome
        this.primaryColor = primaryColor
        this.secondaryColor = secondaryColor
        this.effects.addAll(effects!!)
        this.effects.add("Strength 2")
    }
}